devito/ir/support/basic.py

```diff
@@ -1013,3 +1030,28 @@ def d_from_access(self, accesses):
         TimedAccess objects.
         """
         return DependenceGroup(self.d_from_access_gen(accesses))
+
+    @memoized_generator
+    def r_gen(self):
+        """
+        Generate the Relations of the Scope.
+        """
+        for f in self.functions:
+            v = self.reads.get(f, []) + self.writes.get(f, [])
+
+            for a0, a1 in product(v, v):
+                if a0 is a1:
+                    continue
+
+                r = Relation(a0, a1)
```

Comment:
Is `Relation` commutative? product is quite brutal (might be ok if small set) so if it's commutative you are better of with two for loop `(i, a0) in enumerate(v), a1 in v[I:-1]`

Reply:
it's not commutative because the distance would flip (positive -> negative or negative -> positive)